analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/EmptyCollectionsShouldNotBeEnumeratedBase.cs

```diff
@@ -205,19 +206,19 @@ private static ProgramState ProcessAddMethod(ProgramState state, IMethodSymbol m
         return state;
     }
 
-    private static CollectionConstraint CollectionCreationConstraint(ProgramState state, IObjectCreationOperationWrapper objectCreation) =>
+    private static CollectionConstraint? CollectionCreationConstraint(ProgramState state, IObjectCreationOperationWrapper objectCreation) =>
         objectCreation.Arguments.SingleOrDefault(x => x.ToArgument().Parameter.Type.DerivesOrImplements(KnownType.System_Collections_IEnumerable)) is { } collectionArgument
             ? state[collectionArgument]?.Constraint<CollectionConstraint>()
             : CollectionConstraint.Empty;
 
-    private static NumberConstraint PropertyReferenceConstraint(ProgramState state, IPropertyReferenceOperationWrapper propertyReference) =>
+    private static NumberConstraint? PropertyReferenceConstraint(ProgramState state, IPropertyReferenceOperationWrapper propertyReference) =>
         propertyReference.Property.Name is nameof(Array.Length) or nameof(List<int>.Count)
             ? SizeConstraint(state, propertyReference.Instance)
             : null;
 
     private static NumberConstraint SizeConstraint(ProgramState state, IOperation instance, bool hasFilteringPredicate = false)
     {
-        if (state.ResolveCapture(instance).TrackedSymbol() is { } symbol && state[symbol]?.Constraint<CollectionConstraint>() is { } collection)
+        if (state.ResolveCaptureAndUnwrapConversion(instance).TrackedSymbol() is { } symbol && state[symbol]?.Constraint<CollectionConstraint>() is { } collection)
```

Comment:
This change needs UT to prove that it's really needed

Reply:
I'm wondering if we will ever need `ResolveCapture` without unwrapping the conversation at all.

Can you try putting your new implementation inside `ResolveCapture` to see how much stuff will break?

Reply:
..in another PR